File: java/src/main/java/io/github/muthuishere/toolnexus/examples/Agent.java
package io.github.muthuishere.toolnexus.examples;

import io.github.muthuishere.toolnexus.HttpTool;
import io.github.muthuishere.toolnexus.LlmClient;
import io.github.muthuishere.toolnexus.Tool;
import io.github.muthuishere.toolnexus.Toolkit;
import io.github.muthuishere.toolnexus.Tools;
import io.github.muthuishere.toolnexus.annotations.Param;
import io.github.muthuishere.toolnexus.annotations.ToolMethod;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Full agent example (mirrors js/examples/agent.ts): MCP tools + skills + a native
 * (@ToolMethod) tool + an HTTP tool, driven by the unified client host loop.
 *
 *   OPENROUTER_API_KEY=... gradle runAgent
 */
public final class Agent {

    /** A plain object with an annotated tool method — the Spring-AI @Tool feel. */
    public static final class MathTools {
        @ToolMethod(name = "add", description = "Add two numbers and return the sum.")
        public String add(@Param(name = "a", description = "first addend") double a,
                          @Param(name = "b", description = "second addend") double b) {
            double sum = a + b;
            // print as an integer when whole, to mirror the JS `${a + b}` string
            if (sum == Math.floor(sum) && !Double.isInfinite(sum)) {
                return String.valueOf((long) sum);
            }
            return String.valueOf(sum);
        }
    }

    public static void main(String[] args) throws Exception {
        Toolkit tk = Toolkit.create(new Toolkit.Options()
                .mcpConfig(Examples.fixture("mcp.json"))
                .skillsDir(Examples.fixture("skills")));

        // a native (annotation) tool — collected via reflection
        for (Tool t : Tools.fromObject(new MathTools())) {
            tk.register(t);
        }

        // an HTTP tool (public test API, no auth)
        HttpTool.Options http = new HttpTool.Options();
        http.name = "get_post";
        http.description = "Fetch a placeholder blog post by id from jsonplaceholder.";
        http.method = "GET";
        http.url = "https://jsonplaceholder.typicode.com/posts/{id}";
        Map<String, Object> idProp = new LinkedHashMap<>();
        idProp.put("type", "number");
        idProp.put("description", "post id 1-100");
        Map<String, Object> props = new LinkedHashMap<>();
        props.put("id", idProp);
        Map<String, Object> schema = new LinkedHashMap<>();
        schema.put("type", "object");
        schema.put("properties", props);
        schema.put("required", List.of("id"));
        schema.put("additionalProperties", false);
        http.inputSchema = schema;
        tk.register(HttpTool.of(http));

        StringBuilder toolsLine = new StringBuilder("Tools: [");
        List<Tool> tools = tk.tools();
        for (int i = 0; i < tools.size(); i++) {
            if (i > 0) toolsLine.append(", ");
            toolsLine.append(tools.get(i).name()).append(" (").append(tools.get(i).source()).append(")");
        }
        toolsLine.append("]");
        System.out.println(toolsLine);

        String key = System.getenv("OPENROUTER_API_KEY");
        if (key == null || key.isEmpty()) {
            System.out.println("\n(no OPENROUTER_API_KEY — skipping live loop)");
            tk.close();
            System.exit(0);
        }

        String model = System.getenv("OPENROUTER_MODEL");
        if (model == null || model.isEmpty()) model = "openai/gpt-4o-mini";

        LlmClient agent = LlmClient.create(new LlmClient.Options()
                .baseUrl("https://openrouter.ai/api/v1")
                .style("openai")
                .model(model)
                .apiKey(key)
                .systemPrompt("You are a precise agent. Use tools to compute and fetch facts."));

        LlmClient.RunResult res = agent.run(
                "What is 21 + 21? Use the add tool. Then fetch post id 1 and tell me its title.", tk);

        StringBuilder calls = new StringBuilder();
        for (int i = 0; i < res.toolCalls.size(); i++) {
            if (i > 0) calls.append(", ");
            calls.append(res.toolCalls.get(i).name);
        }
        System.out.println("\nTool calls: [" + calls + "]");
        System.out.println("\nFINAL:\n" + res.text);

        tk.close();
        if (!res.text.contains("42")) {
            System.err.println("❌ expected 42 in answer");
            System.exit(1);
        }
        System.out.println("\n✅ Java agent loop (native + http + mcp + skills) OK");
        System.exit(0);
    }
}

File: elixir/examples/agent.exs
# Full agent example: MCP tools + skills + a native tool + an HTTP tool, driven by
# the unified client host loop. Mirrors js/examples/agent.ts and python/examples/agent.py.
#
# Run from the elixir/ directory (deps fetched):
#     mix run examples/agent.exs                          # tools list only, exits cleanly
#     OPENROUTER_API_KEY=... mix run examples/agent.exs   # runs the live loop
#
# The `everything` MCP server needs `npx` on PATH. The API key is read from the
# environment and never printed.

here = Path.dirname(__ENV__.file)
examples = Path.expand(Path.join([here, "..", "..", "examples"]))

{:ok, tk} =
  Toolnexus.create_toolkit(
    mcp_config: Path.join(examples, "mcp.json"),
    skills_dir: Path.join(examples, "skills")
  )

# a native (annotation) tool
add =
  Toolnexus.define_tool(
    name: "add",
    description: "Add two numbers and return the sum.",
    input_schema: %{
      "type" => "object",
      "properties" => %{"a" => %{"type" => "number"}, "b" => %{"type" => "number"}},
      "required" => ["a", "b"]
    },
    execute: fn %{"a" => a, "b" => b}, _ctx -> to_string(a + b) end
  )

# an HTTP tool (public test API, no auth)
get_post =
  Toolnexus.Http.tool(
    name: "get_post",
    description: "Fetch a placeholder blog post by id from jsonplaceholder.",
    method: "GET",
    url: "https://jsonplaceholder.typicode.com/posts/{id}",
    input_schema: %{
      "type" => "object",
      "properties" => %{"id" => %{"type" => "integer"}},
      "required" => ["id"]
    }
  )

tk = Toolnexus.Toolkit.register(tk, [add, get_post])

IO.puts("MCP status: #{inspect(Toolnexus.Toolkit.mcp_status(tk))}")
IO.puts("Tools (#{length(Toolnexus.Toolkit.tools(tk))}):")

for t <- Toolnexus.Toolkit.tools(tk), do: IO.puts("  - #{t.name} (#{t.source})")

case System.get_env("OPENROUTER_API_KEY") do
  nil ->
    IO.puts("\n(no OPENROUTER_API_KEY set — skipping the live loop; tools listed above)")

  key ->
    client =
      Toolnexus.Client.create(
        base_url: "https://openrouter.ai/api/v1",
        style: "openai",
        model: System.get_env("OPENROUTER_MODEL") || "openai/gpt-4o-mini",
        api_key: key
      )

    result =
      Toolnexus.Client.run(
        client,
        "Use the add tool to add 21 and 21, then tell me the result.",
        tk
      )

    IO.puts("\nAgent answer:\n#{result.text}")
end

Toolnexus.Toolkit.close(tk)
